File: python/tests/end_to_end/end_to_end_test.py
import os
import sys
import json
import io

# Add parent directory to sys.path for imports
sys.path.append(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from assembler import assemble
from emu.cpu import CPU
import tests.tester as tester

def e2e_test(name, data):
    # Pipeline
    program = assemble(data["input"])
    cpu = CPU()
    cpu.load_program(program)

    output_buffer = io.StringIO()
    original_stdout = sys.stdout
    sys.stdout = output_buffer

    cpu.run(data.get("max_cycles", 1000))
    sys.stdout = original_stdout
    output = output_buffer.getvalue()

    success = True

    # Check Output
    if "expected_output" in data and output != data["expected_output"]:
        print(f"Output mismatch in {name}: Expected {repr(data['expected_output'])}, got {repr(output)}")
        success = False

    # Check Memory
    if "expected_memory" in data:
        for addr_str, expected_val in data["expected_memory"].items():
            addr = int(addr_str)
            if cpu.memory[addr] != expected_val:
                print(f"Memory mismatch in {name} at address {addr}: Expected {expected_val}, got {cpu.memory[addr]}")
                success = False

    # Check PC
    if "expected_pc" in data and cpu.pc != data.get("expected_pc"):
        print(f"PC mismatch in {name}: Expected {data['expected_pc']}, got {cpu.pc}")
        success = False

    return success

if __name__ == "__main__":
    with open(os.path.join(os.path.dirname(__file__), "end_to_end_data.json"), "r") as f:
        test_data = json.load(f)
    tester.run("End-to-End Tests", test_data, e2e_test)

File: js/tests/end_to_end/end_to_end_test.js
const assembler = require('../../assembler');
const cpu = require('../../emu/cpu');
const tester = require('../tester');
const testData = require('./end_to_end_data.json');

const e2eTest = (name, data) => {
    // Pipeline
    const program = assembler.assemble(data.input);
    cpu.load_program(program);

    let outputBuffer = "";
    const originalStdoutWrite = process.stdout.write;
    process.stdout.write = (str) => {
        outputBuffer += str;
        return true;
    };

    cpu.run(data.max_cycles || 1000);
    process.stdout.write = originalStdoutWrite;

    let success = true;

    // Check Output
    if (data.expected_output !== undefined && outputBuffer !== data.expected_output) {
        console.log(`Output mismatch in ${name}: Expected ${JSON.stringify(data.expected_output)}, got ${JSON.stringify(outputBuffer)}`);
        success = false;
    }

    // Check Memory
    if (data.expected_memory) {
        for (const addr in data.expected_memory) {
            const expectedVal = data.expected_memory[addr];
            if (cpu.memory[addr] !== expectedVal) {
                console.log(`Memory mismatch in ${name} at address ${addr}: Expected ${expectedVal}, got ${cpu.memory[addr]}`);
                success = false;
            }
        }
    }

    // Check PC
    if (data.expected_pc !== undefined && cpu.pc !== data.expected_pc) {
        console.log(`PC mismatch in ${name}: Expected ${data.expected_pc}, got ${cpu.pc}`);
        success = false;
    }

    return success;
};

tester.run("End-to-End Tests", testData, e2eTest);
